helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java

```diff
@@ -145,9 +145,21 @@ public boolean finishJob(String job) {
           String.format("Job: %s has either finished already, never been scheduled, or been removed from DAG", job));
     }
     // Add finished job's successors to ready-list
+
+    // If it is a jobQueue, there should be a check to make sure that the a job has not been added
+    // to the _readyJobList multiple times. This check is necessary because once the controller
+    // switch happens, the _readyJobList and _inflightJobList will be created from scratch. In this
+    // case, since there might be many jobs that have been finished before, we do not want to have a
+    // job several times to the _readyJobList. If _readyJobList has multiple instances of the same
+    // job, it can compromise the functionality of the parallel jobs.
```

Comment:
Have a follow up question. What do you mean the job gets added several times to the readyJobList?

If there's a controller switch, we'll be creating readyJobList/inflightJobList from scratch, and add everything from the beginning (or the first available job from the step-by-step topological sort). Then as long as the controller doesn't switch, it only gets added once, right?

In other words, for each controller should add a job exactly once. Is that correct?

Also even though a job might get added to readyJobList multiple times (exactly once in different controllers), our scheduling logic just skips it I believe (if jobState is complete/progress we just skip?). Could you help me understand how this could help? It seems that although this wouldn't hurt to do...

Reply:
It is very hard to explain. I will do my best to explain it here. (If it is still not clear, I suggest you to just run the test I added to this PR and it will be very clear to you.)
Explanation: 
Here is the _readyJobList contents after the controller switch in my test (without this fix):

> [testQueueParallelJobs_JOB0, testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB1]
> [testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3]
> [testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3]
> [testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB3]
> [testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB5]
> [testQueueParallelJobs_JOB6]
> 

We end up having jut JOB4 in the DAG and the reason is because once each job is finished, we add new job to the _readyList (_readyList is updated in finishJob method). If there is only one Job in the _readyList, we will wait until 4 is finished and add JOB5 and so on. In this case, we will not run the whole queue in parallel. 

Also interesting to note here that even before controller swicth here is _readyList which is also wrong and we see two JOBs with the same name in the queue. However, this is not as critical as the the one I explained above:

> [testQueueParallelJobs_JOB0, testQueueParallelJobs_JOB1]
> [testQueueParallelJobs_JOB1]
> [testQueueParallelJobs_JOB0, testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB3]
> [testQueueParallelJobs_JOB3]

Reply:
Please note that if we have a lot of jobs in the queue and restart the controller, if we have for example 1000 jobs already in the queue and completed, how many times we have to loop through each job for the first pipeline. It will make the controller very slow (for the fist pipeline) because each job will go to _readyList multiple times.

Reply:
With this fix, this is the _readyList after controller switch:
[testQueueParallelJobs_JOB0, testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
[testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3]
[testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4]
[testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB5]
[testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB5, testQueueParallelJobs_JOB6]
[testQueueParallelJobs_JOB5, testQueueParallelJobs_JOB6]
[testQueueParallelJobs_JOB6]

Reply:
Without the fix:
Please note the we add job to the _readyJob list once we call finishJob (once a job is completed).
Let's say controller switch happened and we started from scratch.
We have [Job0, Job1 , Job2] in the _readyList.
To schedule the jobs, we get next job from readyList (Job0), see that this job is completed, we add the children of Job0 which is Job1. Hence the _readyList will be [Job1, Job2, Job1]. That is why we have same job added to the _readyList.

In conclusion, if a job is finished and controller adds a wrong job, the _readyJobList will fall behind. For example in my test, we add Job4 multiple times. Since Job4 goes to in-progress state and it takes some time to finish, we do not add new job (i.e. job5) until Job4 is finished and hence we do not run jobs in parallel if this happens.

Reply:
@alirezazamani Thank for explaining with concrete examples. So the root cause seems to be that we do not do if (!_readyJobList.contains(nextJob)) when we add the jobs to the ready list.

Good find and good fix! :)